core/sched.cc

```diff
@@ -530,6 +554,16 @@ void cpu::enqueue(thread& t)
     runqueue.insert_equal(t);
 }
 
+// When the run queue has several threads with equal thread_runtime_compare,
+// enqueue() puts a thread after its equals, while enqueue_first_equal()
+// puts it before its equals. The distinction is mostly interesting for real-
+// time priority threads.
+void cpu::enqueue_first_equal(thread& t)
+{
+    trace_sched_queue(&t);
+    runqueue.insert_before(runqueue.lower_bound(t), t);
```

Comment:
I assume this behaves the way we want it - "1st among the equals ... in terms of priority" given the modified `thread_runtime_compare`? And if all threads have priority 0, the `enqueue_first_equal` would behave like `enqueue`?

Reply:
enqueue_first_equal() is exactly the same as the old enqueue(), it just says that if there are already several threads with exactly the same "runtime_compare", the thread gets put *before* those equal items in the queue, instead of after them as usual. 
Note that the only real case of exact equality of runtime_compare is in real-time threads which have the same real-time priority (for non-realtime threads, the chance that the specific runtime is identical is very small and not interesting).
We need to use it in exactly one place: When a thread of realtime priority 1 is preempted by a thread of realtime priority 2 (for example), we need to put the thread with priority 1 back into the runqueue, and we need to put it **first** (not last) so that when the thread of priority 2 is done, we go back to the original thread with priority 1 - not the next one. This is a requirement of POSIX.